Upgrade Google Auth

dextermb opened the pull request:
Resolves #25

Upgrades from using `gapi.auth2` to `google.accounts.id` (https://developers.google.com/identity/gsi/web). 

<img width="268" alt="Screenshot 2022-07-22 at 16 33 00" src="https://user-images.githubusercontent.com/6137789/180473664-a86066c2-4520-40ab-94c7-8c1e0610ea7d.png">

This change, however, is **breaking** as the outputs from the events are different.

* `auth-success` now returns an object with user properties:
  * `name`
  * `email`
  * `picture`
* `auth-failure` now returns a basic string
* `init-error` is no longer used

dextermb committed:
fix(dependency): Upgrade svelte kit

dextermb committed:
feature(gauth): Upgrade to use GSI

dextermb commented:
Once people are happy with the implementation then I will look at updating the documentation

antony commented on package.json:
```diff
@@ -1,26 +1,27 @@
+    "vite": "^3.0.0"
   },
-  "type": "module",
-  "volta": {
```
Why is volta gone?

dextermb replied:
Removed as part of the upgrade to SvelteKit, I can add it back with a newer version of node

dextermb replied:
Added back with node `16.15.0` ✅

antony commented on src/lib/google-auth/GoogleAuth.svelte:
```diff
@@ -1,4 +1,8 @@
+<button
+  {disabled}
+  use:action
+  on:click={handleClick}
```
We should call event handler functions after what they do, not what events they handle

dextermb replied:
The click here triggers the Google Sign In popup. Would the function name `triggerPopup` work for you?

dextermb replied:
Renamed to `triggerPrompt` ✅

antony commented on src/lib/google-auth/GoogleAuth.svelte:
```diff
@@ -73,22 +78,27 @@
+
+    const { name, email, picture } = decode(credential)
+
+    dispatch('auth-success', { user: { name, email, picture } })
```
How do we get their Google id to write to their profile?

dextermb replied:
I believe we can use the `sub` (subject) property of the JWT token:
https://developers.google.com/identity/gsi/web/reference/js-reference#credential

dextermb replied:
Added `sub` renamed as `id` on output of `auth-success` ✅

antony reviewed (changes requested):
Looks good, a few improvements for readability and a question

antony commented:
Note that we don't validate the jwt signature, so it might be better to get rid of the new dependency and just bade64 decode the unencrypted part of the jwt

https://stackoverflow.com/questions/38552003/how-to-decode-jwt-token-in-javascript-without-using-a-library

dextermb commented:
> Note that we don't validate the jwt signature, so it might be better to get rid of the new dependency and just bade64 decode the unencrypted part of the jwt
> 
> https://stackoverflow.com/questions/38552003/how-to-decode-jwt-token-in-javascript-without-using-a-library

Certainly if you're happy with having the suggested browser function in the codebase. I'll update this

dextermb committed:
fix(google-auth): Update based on PR feedback

dextermb commented:
> Certainly if you're happy with having the suggested browser function in the codebase. I'll update this

Removed `jwt-decode` dependency and added the answer-marked implementation from stackoverflow ✅

dextermb committed:
Update README

antony commented:
lets pnpm link this to the app to try it first before we merge and release.

dextermb committed:
feat(google-auth): Expose JWT token

dextermb commented:
> lets pnpm link this to the app to try it first before we merge and release.

👋🏻 Hey, just getting back to this today. I've done a decent amount of testing and reading through our internal use-cases.

@antony with minor changes to our login screen I've got this working as we would expect. I've exposed the whole token through the `auth-success` callback.

It's worth noting the GSI does not like running on non-https domains, other than localhost, so this will throw some complications for us and potentially other development workflows.

dextermb commented:
I've also noticed that `button` is now being respected on our login/sign up screens- e.g. not `width: 100%` by default which is correct. Potentially something else has changed in our platform but to match the same styles as previously seen I updated the styles:

```css
  :global(.social-auth-button button) {
    width: 100%;
    font-family: 'Cabin', sans-serif;
    border-radius: 5px;
  }
```

_But I don't think this is anything to do with this pull request as looking at the code I have not changed anything relating to styles._

TheOnlyWayUp commented:
Thanks for the PR! I installed from this commit and importing failed, but copying over the GoogleAuth.svelte file and using it like any other component seems to work

:)

antony commented:
lets also consider this https://github.com/beyonk-adventures/svelte-social-auth/pull/14 when we merge
